src/Polysemy/Cont/Internal.hs

```diff
@@ -95,4 +103,155 @@ embedSem = liftSem . weave (pure ()) (pure . join) inspectSem
+  'runContViaFreshInCWeave', so I need to think it through some more.
+-}
+-- | Intermediary monadic interpretation used for running 'runContViaFresh'.
+-- See source for a discussion on how this works.
+runContViaFreshInC :: forall uniq s r a
+                    . (Member (Fresh uniq) r, Eq uniq)
+                   => Sem (Cont (ViaFreshRef uniq) ': r) a
+                   -> ContT s (Sem (Error (uniq, Any) ': r)) a
+runContViaFreshInC = usingSem $ \u -> ContT $ \c ->
+  case decomp u of
+    Right (Weaving e s wv ex _) ->
+      case e of
+        Subst main cn -> do
+          ref <- fresh
+          let
+            main' = runContViaFreshInC . wv . fmap main . (<$ s)
+            cn'   = runContViaFreshInC . wv . fmap cn . (<$ s)
+            loop act =
+              runContT (ex <$> act) c `catch` \ x@(ref', a') -> do
+                if ref == ref' then
+                  loop (cn' $ unsafeCoerce a')
+                else
+                  throw x
+          loop $ main' $ ViaFreshRef (\a -> (ref, unsafeCoerce a))
+        Jump ref a -> throw (getBacktrackException ref a)
+    Left g -> do
+      ResAndHandler a rc <- liftSem $
+        weave
+          (ResAndHandler @uniq @r () throw)
+          -- TODO(KingoftheHomeless): is this the distributive law we want?
+          (\(ResAndHandler a rc) ->
+            runContT
+              (runContViaFreshInCWeave a)
+              (\x -> pure $
+                ResAndHandler
+                  x
+                  (rc >=> (`runContT` pure) . runContViaFreshInC)
+              )
+          )
+          (Just . getResult)
+          (weaken g)
+      let loop x = c x `catch` (rc >=> loop)
+      loop a
+
+-- | A variant of 'runContViaFreshInC' which it uses when weaving other effects through.
+runContViaFreshInCWeave :: forall uniq s r a
+                         . (Member (Fresh uniq) r, Eq uniq)
+                        => Sem (Cont (ViaFreshRef uniq) ': r) a
+                        -> ContT (ContFreshState uniq r s)
+                            (Sem (Error (uniq, Any) ': r))
+                            a
+runContViaFreshInCWeave = usingSem $ \u -> ContT $ \c ->
+  case decomp u of
+    Right (Weaving e s wv ex _) ->
+      case e of
+        Subst main cn -> do
+          ref <- fresh
+          let
+            -- TODO(KingoftheHomeless): runContViaFreshInC?
+            main' = runContViaFreshInCWeave . wv . fmap main . (<$ s)
+            cn'   = runContViaFreshInCWeave . wv . fmap cn . (<$ s)
+            loop act =
+              runContT (ex <$> act) c `catch` \ x@(ref', a') -> do
+                if ref == ref' then
+                  loop (cn' $ unsafeCoerce a')
+                else
+                  throw x
+          ResAndHandler res h <-
+            loop $ main' $ ViaFreshRef (\a -> (ref, unsafeCoerce a))
+          return $ ResAndHandler res
+              -- TODO(KingoftheHomeless): This handler is dubious.
```

Comment:
How so?

Reply:
I mostly built these handlers by kludging the free variables together until they type-checked, without any greater regard if the way I constructed them was right. It seems to work, but since I constructed them haphazardly, they're dubious.

Reply:
that's how i built the whole library ;)